sim-lib/src/lib.rs

```diff
@@ -429,44 +444,54 @@ async fn consume_events(
         match event {
             SimulationEvent::SendPayment(dest, amt_msat) => {
                 let mut node = node.lock().await;
-                let payment = node.send_payment(dest, amt_msat);
 
-                match payment.await {
+                let mut payment = Payment {
+                    source: node.get_info().pubkey,
+                    hash: None,
+                    amount_msat: amt_msat,
+                    destination: dest,
+                    dispatch_time: SystemTime::now(),
+                };
+
+                let outcome = match node.send_payment(dest, amt_msat).await {
                     Ok(payment_hash) => {
                         log::debug!(
                             "Send payment: {} -> {}: ({}).",
                             node_id,
                             dest,
                             hex::encode(payment_hash.0)
                         );
+                        // We need to track the payment outcome using the payment hash that we have received.
+                        payment.hash = Some(payment_hash);
+                        SimulationOutput::SendPaymentSuccess(payment)
+                    }
+                    Err(e) => {
+                        log::error!("Error while sending payment {} -> {}.", node_id, dest);
 
-                        log::debug!("Sending output for {}.", hex::encode(payment_hash.0));
-                        let output = SimulationOutput::PaymentSent(DispatchedPayment {
-                            source: node.get_info().pubkey,
-                            hash: payment_hash,
-                            amount_msat: amt_msat,
-                            destination: dest,
-                            dispatch_time: SystemTime::now(),
-                        });
-
-                        match sender.send(output).await {
-                            Ok(_) => {}
-                            Err(e) => {
-                                log::error!("Error sending simulation output: {:?}.", e);
+                        match e {
+                            LightningError::PermanentError(s) => {
+                                log::error!("Simulation terminated with error: {s}.");
+                                shutdown.trigger();
                                 break;
                             }
+                            _ => SimulationOutput::SendPaymentFailure(
+                                payment,
+                                PaymentResult {
+                                    htlc_count: 0,
+                                    payment_outcome: PaymentOutcome::NotDispatched,
+                                },
```

Comment:
I think it may be worth creating a couple of helper constructors to reduce the boilerplate here:

```
PaymentResult::not_dispatched();
PaymentResult::tracked_payment_failed();
```